src/status_im2/contexts/chat/lightbox/zoomable_image/view.cljs

```diff
@@ -0,0 +1,313 @@
+  ([value duration]
+   (reanimated/with-timing-duration value duration)))
+
+(defn set-val-decay
+  [animation value bounds]
+  (reanimated/animate-shared-value-with-decay animation value bounds))
+
+;; MATH
+(defn get-max-offset
+  [size screen-size scale]
+  (/ (/ (- (* size scale) screen-size) 2) scale))
+
+(defn get-scale-ratio
+  [new-scale saved-scale]
+  (/ (- (dec new-scale) (dec saved-scale)) new-scale))
+
+(defn get-current-center
+  [size scaled-size offset]
+  (- (+ (/ (- size scaled-size) 2) (/ scaled-size 2)) offset))
+
+;; 5 Gestures: tap, double-tap, pinch, pan-x, pan-y
+(defn tap-gesture
+  [on-tap]
+  (->
+    (.Tap gesture/gesture)
+    (.onStart #(on-tap))))
+
+(defn double-tap-gesture
+  [{:keys [width height screen-height]}
+   {:keys [scale pan-x pan-x-start pan-y pan-y-start]}
+   {:keys [y-threshold-scale]}
+   rescale]
+  (->
+    (.Tap gesture/gesture)
+    (.numberOfTaps 2)
+    (.onStart (fn [evt]
+                ;; Scale to x2
+                (if (= (get-val scale) min-scale)
+                  ;; Find translate-x value
+                  (let [focal-x     (- (/ width 2) (.-x evt))
+                        max-pan-x   (get-max-offset width width double-tap-scale)
+                        translate-x (min (Math/abs focal-x) max-pan-x)
+                        translate-x (if (neg? focal-x) (- translate-x) translate-x)
+                        ;; Find translate-y value
+                        focal-y     (- (/ height 2) (.-y evt))
+                        max-pan-y   (get-max-offset height screen-height double-tap-scale)
+                        translate-y (min (Math/abs focal-y) max-pan-y)
+                        translate-y (if (neg? focal-y) (- translate-y) translate-y)]
+                    ;; apply animations
+                    (set-val pan-x (timing translate-x))
+                    (set-val pan-x-start (timing translate-x))
+                    ;; animate y position only if the double-tap-scale exceeds the threshold
+                    (when (> double-tap-scale y-threshold-scale)
+                      (set-val pan-y (timing translate-y))
+                      (set-val pan-y-start (timing translate-y)))
+                    (rescale double-tap-scale))
+                  ;; Otherwise scale back to x1
+                  (rescale min-scale))))))
+
+(defn pinch-gesture
+  [{:keys [width height]}
+   {:keys [saved-scale scale pinch-x pinch-y pinch-x-start pinch-y-start pan-x-start pan-y-start]}
+   {:keys [pan-x-enabled? pan-y-enabled? x-threshold-scale y-threshold-scale focal-x focal-y]}
+   rescale]
+  (->
+    (.Pinch gesture/gesture)
+    (.onStart (fn [evt]
+                ;; Read the x and y touch positions on the screen
+                (reset! focal-x (.-focalX evt))
+                (reset! focal-y (.-focalY evt))))
+    (.onUpdate (fn [evt]
+                 ;; First, find the scale ration
+                 (let [new-scale     (* (.-scale evt) (get-val saved-scale))
```

Comment:
personally, I prefer code without such comments. You can put this into helper functions which explains what each part is doing and then you should naturally need less comments 

```
(defn find-pinch-x [width saved-scale pinch-x-start pan-x-start focal-x scale-ratio]
  (let [scaled-width  (/ width (get-val saved-scale))
        x-start       (+ (get-val pinch-x-start) (get-val pan-x-start))
        center-x      (get-current-center width scaled-width x-start)]
    (* (- center-x @focal-x) scale-ratio)))

(defn find-pinch-y [height pinch-y-start scale pan-y-start focal-y scale-ratio]
  (let [scaled-height (/ height (get-val scale))
        y-start       (+ (get-val pinch-y-start) (get-val pan-y-start))
        center-y      (get-current-center height scaled-height y-start)]
    (* (- center-y @focal-y) scale-ratio)))

(defn pinch-gesture
  [{:keys [width height]}
   {:keys [saved-scale scale pinch-x pinch-y pinch-x-start pinch-y-start pan-x-start pan-y-start]}
   {:keys [pan-x-enabled? pan-y-enabled? x-threshold-scale y-threshold-scale focal-x focal-y]}
   rescale]
  (->
   (.Pinch gesture/gesture)
   (.onStart (fn [evt]
                ;; Read the x and y touch positions on the screen
               (reset! focal-x (.-focalX evt))
               (reset! focal-y (.-focalY evt))))
   (.onUpdate (fn [evt]
                (let [new-scale     (* (.-scale evt) (get-val saved-scale))
                      scale-ratio   (get-scale-ratio new-scale (get-val saved-scale))
                      new-pinch-x   (find-pinch-x width saved-scale pinch-x-start pan-x-start focal-x scale-ratio) 
                      new-pinch-y    (find-pinch-y height pinch-y-start scale pan-y-start focal-y scale-ratio)]
                  (set-val pinch-x (+ new-pinch-x (get-val pinch-x-start)))
                  (set-val pinch-y (+ new-pinch-y (get-val pinch-y-start)))
                  (set-val scale new-scale))))))
```

Reply:
same for `onStart` method, could be
```
(defn read-x-and-y-on-screen [focal-x focal-y]
  (fn [evt]
                ;; Read the x and y touch positions on the screen
    (reset! focal-x (.-focalX evt))
    (reset! focal-y (.-focalY evt))))

(defn pinch-gesture
  [{:keys [width height]}
   {:keys [saved-scale scale pinch-x pinch-y pinch-x-start pinch-y-start pan-x-start pan-y-start]}
   {:keys [pan-x-enabled? pan-y-enabled? x-threshold-scale y-threshold-scale focal-x focal-y]}
   rescale]
  (->
   (.Pinch gesture/gesture)
   (.onStart (read-x-and-y-on-screen focal-x focal-y))
   ```

Reply:
@J-Son89 fixed it.